### ISSUE TITLE
Compilation of library imports into LLVM IR/Bitcode

### ISSUE THREAD

**Issue by lung21** [sbip-sg/tsll#4]
#### Motivation

It is not enough to run a program in LLVM IR or Bitcode by itself, because the program may have been written with other library functions, classes, etc. In order for us to run it with `lli`, an LLVM interpreter, we need a way to compile and link these libraries used in that program.

#### Example
If we have the following program where `console.log` is from another library, we should compile its definition into LLVM IR, or link the compiled library with `lli`.
```
let str = 'string';
let a = 3;
let b = 4 + 1;
let c = a + 1;
function testCase1(a: number): void {
    return;
}

a = b;

console.log('print something meaningful');
```

#### Expected output
Then, we should receive the message as follows if we run the program above.
```
print something meaningful
```

**Comment by taquangtrung** [comment]
Thanks Lung! I see the problem here. The compilation of JS/TS library might be a heavy workload.

How about we compile the LLVM bitcode back to Javascript using Emscript? 
- https://dl.acm.org/doi/abs/10.1145/2048147.2048224
- https://emscripten.org/docs/introducing_emscripten/about_emscripten.html

Then we don't need to us `lli` to run the bitcode, just run the original Typescript programs and the Javascript programs obtained from Tsll + Emscripten compilation and compare their result.

**Comment by lung21** [comment]
> Thanks Lung! I see the problem here. The compilation of JS/TS library might be a heavy workload.
> 
> How about we compile the LLVM bitcode back to Javascript using Emscript?
> 
> * https://dl.acm.org/doi/abs/10.1145/2048147.2048224
> * https://emscripten.org/docs/introducing_emscripten/about_emscripten.html
> 
> Then we don't need to use `lli` to run the bitcode, just run the original Typescript programs and the Javascript programs obtained from Tsll + Emscripten compilation and compare their result.

@taquangtrung 

I think that declarations of any JS/TS library still have to be imported into a TS program we are compiling to LLVM IR. The reason is that it would use an imported class and possibly its member functions in the program.

**Comment by taquangtrung** [comment]
I agree. For compilation backward from LLVM IR to Javascript, we need to declare the used functions/libraries.

But we only need to provide the declarations, not the definitions, right?

**Comment by lung21** [comment]
Exactly. I'm thinking over how we can implement our idea.
